### ISSUE TITLE
selftests: diag: failing: found msk after flush while expected none

### ISSUE THREAD

**Issue by matttbe** [multipath-tcp/mptcp_net-next#339]
With this patch provided by @pabeni

```diff
diff --git a/include/net/sock.h b/include/net/sock.h
index dcd72e6285b2..23425610177f 100644
--- a/include/net/sock.h
+++ b/include/net/sock.h
@@ -1534,11 +1534,8 @@ struct prot_inuse {
        int val[PROTO_INUSE_NR];
 };
 
-static inline void sock_prot_inuse_add(const struct net *net,
-                                      const struct proto *prot, int val)
-{
-       this_cpu_add(net->core.prot_inuse->val[prot->inuse_idx], val);
-}
+void sock_prot_inuse_add(const struct net *net,
+                        const struct proto *prot, int val);
 
 static inline void sock_inuse_add(const struct net *net, int val)
 {
diff --git a/net/core/sock.c b/net/core/sock.c
index f954d5893e79..7a884b8b323e 100644
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -3744,6 +3744,16 @@ int sock_inuse_get(struct net *net)
 
 EXPORT_SYMBOL_GPL(sock_inuse_get);
 
+void sock_prot_inuse_add(const struct net *net,
+                                      const struct proto *prot, int val)
+{
+       static int cnt;
+
+       this_cpu_add(net->core.prot_inuse->val[prot->inuse_idx], val);
+       cnt = this_cpu_read(net->core.prot_inuse->val[prot->inuse_idx]);
+}
+EXPORT_SYMBOL(sock_prot_inuse_add);
+
 static int __net_init sock_inuse_init_net(struct net *net)
 {
        net->core.prot_inuse = alloc_percpu(struct prot_inuse);
diff --git a/tools/testing/selftests/net/mptcp/diag.sh b/tools/testing/selftests/net/mptcp/diag.sh
index ef628b16fe9b..69d203704f3e 100755
--- a/tools/testing/selftests/net/mptcp/diag.sh
+++ b/tools/testing/selftests/net/mptcp/diag.sh
@@ -178,6 +178,18 @@ chk_msk_inuse()
        done
 
        __chk_nr get_msk_inuse $expected $*
+       if [ $((ret + 1)) == $test_cnt ]; then
+               ip netns exec $ns grep MPTCP /proc/net/protocols
+               ss -pinmHMN $ns
+               ip netns exec $ns nstat -as Tcp* MPTcp*
+
+               echo "let mptcp timeout elapse"
+               sleep 70
+               ss -pinmHMN $ns
+               ip netns exec $ns nstat -as Tcp* MPTcp*
+               ip netns exec $ns grep MPTCP /proc/net/protocols
+               exit $ret
+       fi
 }
 
 # $1: ns, $2: port
```

I got:

```
++ perf record -ag -e probe:sock_prot_inuse_add -e probe:__mptcp_destroy_sock ./diag.sh
no msk on netns creation                          [  ok  ]
listen match for dport 10000                      [  ok  ]
listen match for sport 10000                      [  ok  ]
listen match for saddr and sport                  [  ok  ]
all listen sockets                                [  ok  ]
after MPC handshake                               [  ok  ]
....chk remote_key                                [  ok  ]
....chk no fallback                               [  ok  ]
....chk 2 msk in use                              [  ok  ]
....chk 0 msk in use after flush                  [  ok  ]
check fallback                                    [  ok  ]
....chk 1 msk in use                              [  ok  ]
....chk 0 msk in use after flush                  [ fail ] expected 0 found 1
MPTCPv6   2936      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
MPTCP     2784      1       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
#kernel
TcpActiveOpens                  2                  0.0
TcpPassiveOpens                 2                  0.0
TcpInSegs                       25                 0.0
TcpOutSegs                      25                 0.0
TcpOutRsts                      1                  0.0
TcpExtTW                        1                  0.0
TcpExtTCPPureAcks               11                 0.0
TcpExtTCPHPAcks                 1                  0.0
TcpExtTCPAbortOnClose           1                  0.0
TcpExtTCPOrigDataSent           8                  0.0
TcpExtTCPDelivered              8                  0.0
MPTcpExtMPCapableSYNRX          1                  0.0
MPTcpExtMPCapableSYNTX          2                  0.0
MPTcpExtMPCapableSYNACKRX       1                  0.0
MPTcpExtMPCapableACKRX          1                  0.0
MPTcpExtMPCapableFallbackSYNACK 1                  0.0
let mptcp timeout elapse
#kernel
TcpActiveOpens                  2                  0.0
TcpPassiveOpens                 2                  0.0
TcpInSegs                       25                 0.0
TcpOutSegs                      25                 0.0
TcpOutRsts                      1                  0.0
TcpExtTW                        1                  0.0
TcpExtTCPPureAcks               11                 0.0
TcpExtTCPHPAcks                 1                  0.0
TcpExtTCPAbortOnClose           1                  0.0
TcpExtTCPOrigDataSent           8                  0.0
TcpExtTCPDelivered              8                  0.0
MPTcpExtMPCapableSYNRX          1                  0.0
MPTcpExtMPCapableSYNTX          2                  0.0
MPTcpExtMPCapableSYNACKRX       1                  0.0
MPTcpExtMPCapableACKRX          1                  0.0
MPTcpExtMPCapableFallbackSYNACK 1                  0.0
MPTCPv6   2936      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
MPTCP     2784      1       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
[ perf record: Woken up 1 times to write data ]
failed to mmap file
[ perf record: Captured and wrote 0.260 MB perf.data ]
++ rc=1
```

So still the same after 70 sec.

Info from Perf coming from:

```
perf probe -a sock_prot_inuse_add
perf probe -a __mptcp_destroy_sock
cd tools/testing/selftests/net/mptcp
run_loop perf record -ag -e probe:sock_prot_inuse_add -e probe:__mptcp_destroy_sock ./diag.sh
perf script > perf.data.txt
```
[perf.data.txt.zip](https://github.com/multipath-tcp/mptcp_net-next/files/10482167/perf.data.txt.zip)

With more probes:

```
perf probe -a sock_prot_inuse_add
perf probe -a __mptcp_destroy_sock
perf probe -k .virtme/build/vmlinux -a 'subflow_state_change sk sk->__sk_common.skc_state'
perf probe -k .virtme/build/vmlinux -a '__mptcp_close:8 msk=sk sk=msk->first sk->__sk_common.skc_state'
cd tools/testing/selftests/net/mptcp
run_loop _tap ${RESULTS_DIR}/perf.tap perf record -ag -e probe:sock_prot_inuse_add -e probe:__mptcp_destroy_sock -e probe:subflow_state_change -e probe:__mptcp_close_L8 -o perf.data ./diag.sh
perf script > perf.data.txt
```

[perf.data_202301241347.zip](https://github.com/multipath-tcp/mptcp_net-next/files/10489795/perf.data_202301241347.zip)

Reproducible using:

```bash
cd <kernel source code>

cat <<'EOF' > .virtme-exec-run
perf probe -a sock_prot_inuse_add
perf probe -a __mptcp_destroy_sock
perf probe -k .virtme/build/vmlinux -a 'subflow_state_change sk sk->__sk_common.skc_state'
perf probe -k .virtme/build/vmlinux -a '__mptcp_close:8 msk=sk sk=msk->first sk->__sk_common.skc_state'
cd tools/testing/selftests/net/mptcp
run_loop _tap ${RESULTS_DIR}/perf.tap perf record -ag -e probe:sock_prot_inuse_add -e probe:__mptcp_destroy_sock -e probe:subflow_state_change -e probe:__mptcp_close_L8 -o perf.data ./diag.sh
perf script > perf.data.txt
EOF

docker run -v "${PWD}:${PWD}:rw" -w "${PWD}" --privileged --rm -it --pull always mptcp/mptcp-upstream-virtme-docker:latest auto-debug
```

**Comment by matttbe** [comment]
@pabeni: I managed to reproduce the same warning but a different issue on top of your patches from #320 and the fix you sent for this #339:

```
# no msk on netns creation                          [  ok  ]
# listen match for dport 10000                      [  ok  ]
# listen match for sport 10000                      [  ok  ]
# listen match for saddr and sport                  [  ok  ]
# all listen sockets                                [  ok  ]
# after MPC handshake                               [  ok  ]
# ....chk remote_key                                [  ok  ]
# ....chk no fallback                               [  ok  ]
# ....chk 2 msk in use                              [  ok  ]
# ....chk 0 msk in use after flush                  [  ok  ]
# check fallback                                    [  ok  ]
# ....chk 1 msk in use                              [  ok  ]
# ....chk 0 msk in use after flush                  [  ok  ]
# many msk socket present                           [  ok  ]
# ....chk many msk in use                           [  ok  ]
# ....chk 0 msk in use after flush                  [ fail ] expected 0 found 2
# MPTCPv6   2936      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# MPTCP     2784      2       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# LAST-ACK 0      0      127.0.0.1:38114 127.0.0.1:10006
#        skmem:(r0,rb131072,t0,tb16384,f0,w0,o0,bl0,d0) subflows_max:2 remote_key token:2a999a38 write_seq:4bbde1055e01f245 snd_una:4bbde1055e01f244 rcv_nxt:d9795a810e41cfd4
# LAST-ACK 0      0      127.0.0.1:36714 127.0.0.1:10066
#        skmem:(r0,rb131072,t0,tb16384,f0,w0,o0,bl0,d0) subflows_max:2 remote_key token:9dec73d1 write_seq:ae1ec9a088d6f32f snd_una:ae1ec9a088d6f32e rcv_nxt:37fd3816906823f6
# #kernel
# TcpActiveOpens                  102                0.0
# TcpPassiveOpens                 102                0.0
# TcpEstabResets                  198                0.0
# TcpInSegs                       1016               0.0
# TcpOutSegs                      1016               0.0
# TcpOutRsts                      110                0.0
# TcpExtTW                        4                  0.0
# TcpExtTCPPureAcks               480                0.0
# TcpExtTCPHPAcks                 1                  0.0
# TcpExtTCPAbortOnData            99                 0.0
# TcpExtTCPOrigDataSent           210                0.0
# TcpExtTCPDelivered              312                0.0
# MPTcpExtMPCapableSYNRX          101                0.0
# MPTcpExtMPCapableSYNTX          102                0.0
# MPTcpExtMPCapableSYNACKRX       101                0.0
# MPTcpExtMPCapableACKRX          101                0.0
# MPTcpExtMPCapableFallbackSYNACK 1                  0.0
# MPTcpExtMPFastcloseTx           99                 0.0
# MPTcpExtMPFastcloseRx           94                 0.0
# MPTcpExtMPRstTx                 99                 0.0
# MPTcpExtMPRstRx                 94                 0.0
# let mptcp timeout elapse
# #kernel
# TcpActiveOpens                  102                0.0
# TcpPassiveOpens                 102                0.0
# TcpEstabResets                  198                0.0
# TcpInSegs                       1016               0.0
# TcpOutSegs                      1016               0.0
# TcpOutRsts                      110                0.0
# TcpExtTW                        4                  0.0
# TcpExtTCPPureAcks               480                0.0
# TcpExtTCPHPAcks                 1                  0.0
# TcpExtTCPAbortOnData            99                 0.0
# TcpExtTCPOrigDataSent           210                0.0
# TcpExtTCPDelivered              312                0.0
# MPTcpExtMPCapableSYNRX          101                0.0
# MPTcpExtMPCapableSYNTX          102                0.0
# MPTcpExtMPCapableSYNACKRX       101                0.0
# MPTcpExtMPCapableACKRX          101                0.0
# MPTcpExtMPCapableFallbackSYNACK 1                  0.0
# MPTcpExtMPFastcloseTx           99                 0.0
# MPTcpExtMPFastcloseRx           94                 0.0
# MPTcpExtMPRstTx                 99                 0.0
# MPTcpExtMPRstRx                 94                 0.0
# MPTCPv6   2936      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# MPTCP     2784      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# [ perf record: Woken up 6 times to write data ]
# failed to mmap file
# [ perf record: Captured and wrote 1.575 MB perf.data ]
++ rc=12
```

Here, we can see the sockets are removed after 70 seconds.

[perf.data_202301251954.zip](https://github.com/multipath-tcp/mptcp_net-next/files/10502633/perf.data_202301251954.zip)

(I guess this is not related to your patches from #320 but a timing issue)

**Comment by pabeni** [comment]
> @pabeni: I managed to reproduce the same warning but a different issue on top of your patches from #320 and the fix you sent for this #339:
[...]
> (I guess this is not related to your patches from #320 but a timing issue)

I think this is instead related to the patches from #320, as I could not reproduce the new issue without the changes from #320.

For sure is a different issue from the one addressed with https://lore.kernel.org/mptcp/411cc0b4-af7f-5019-b2ac-a7361f3dcaa9@tessares.net/T/#t, as here no fallback socket is involved.

I think the fastclose is not generated at close time for some edgy scenarios

**Comment by matttbe** [comment]
After 5 hours of tests and 757 attempts, I managed to reproduce it on top of the `export` branch + the patch you sent for this issue (#339 - `mptcp: do not propagate fallback subflow status on error`) and your debug commit from above:

```
# no msk on netns creation                          [  ok  ]
# listen match for dport 10000                      [  ok  ]
# listen match for sport 10000                      [  ok  ]
# listen match for saddr and sport                  [  ok  ]
# all listen sockets                                [  ok  ]
# after MPC handshake                               [  ok  ]
# ....chk remote_key                                [  ok  ]
# ....chk no fallback                               [  ok  ]
# ....chk 2 msk in use                              [  ok  ]
# ....chk 0 msk in use after flush                  [  ok  ]
# check fallback                                    [  ok  ]
# ....chk 1 msk in use                              [  ok  ]
# ....chk 0 msk in use after flush                  [  ok  ]
# many msk socket present                           [  ok  ]
# ....chk many msk in use                           [  ok  ]
# ....chk 0 msk in use after flush                  [ fail ] expected 0 found 2
# MPTCPv6   2936      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# MPTCP     2784      2       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# LAST-ACK 0      0      127.0.0.1:38114 127.0.0.1:10006
#        skmem:(r0,rb131072,t0,tb16384,f0,w0,o0,bl0,d0) subflows_max:2 remote_key token:2a999a38 write_seq:4bbde1055e01f245 snd_una:4bbde1055e01f244 rcv_nxt:d9795a810e41cfd4
# LAST-ACK 0      0      127.0.0.1:36714 127.0.0.1:10066
#        skmem:(r0,rb131072,t0,tb16384,f0,w0,o0,bl0,d0) subflows_max:2 remote_key token:9dec73d1 write_seq:ae1ec9a088d6f32f snd_una:ae1ec9a088d6f32e rcv_nxt:37fd3816906823f6
# #kernel
# TcpActiveOpens                  102                0.0
# TcpPassiveOpens                 102                0.0
# TcpEstabResets                  198                0.0
# TcpInSegs                       1016               0.0
# TcpOutSegs                      1016               0.0
# TcpOutRsts                      110                0.0
# TcpExtTW                        4                  0.0
# TcpExtTCPPureAcks               480                0.0
# TcpExtTCPHPAcks                 1                  0.0
# TcpExtTCPAbortOnData            99                 0.0
# TcpExtTCPOrigDataSent           210                0.0
# TcpExtTCPDelivered              312                0.0
# MPTcpExtMPCapableSYNRX          101                0.0
# MPTcpExtMPCapableSYNTX          102                0.0
# MPTcpExtMPCapableSYNACKRX       101                0.0
# MPTcpExtMPCapableACKRX          101                0.0
# MPTcpExtMPCapableFallbackSYNACK 1                  0.0
# MPTcpExtMPFastcloseTx           99                 0.0
# MPTcpExtMPFastcloseRx           94                 0.0
# MPTcpExtMPRstTx                 99                 0.0
# MPTcpExtMPRstRx                 94                 0.0
# let mptcp timeout elapse
# #kernel
# TcpActiveOpens                  102                0.0
# TcpPassiveOpens                 102                0.0
# TcpEstabResets                  198                0.0
# TcpInSegs                       1016               0.0
# TcpOutSegs                      1016               0.0
# TcpOutRsts                      110                0.0
# TcpExtTW                        4                  0.0
# TcpExtTCPPureAcks               480                0.0
# TcpExtTCPHPAcks                 1                  0.0
# TcpExtTCPAbortOnData            99                 0.0
# TcpExtTCPOrigDataSent           210                0.0
# TcpExtTCPDelivered              312                0.0
# MPTcpExtMPCapableSYNRX          101                0.0
# MPTcpExtMPCapableSYNTX          102                0.0
# MPTcpExtMPCapableSYNACKRX       101                0.0
# MPTcpExtMPCapableACKRX          101                0.0
# MPTcpExtMPCapableFallbackSYNACK 1                  0.0
# MPTcpExtMPFastcloseTx           99                 0.0
# MPTcpExtMPFastcloseRx           94                 0.0
# MPTcpExtMPRstTx                 99                 0.0
# MPTcpExtMPRstRx                 94                 0.0
# MPTCPv6   2936      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# MPTCP     2784      0       0   no       0   no   kernel      y  y  y  y  y  y  y  y  y  y  y  y  n  n  n  y  y  y  n
# [ perf record: Woken up 6 times to write data ]
# failed to mmap file
# [ perf record: Captured and wrote 1.575 MB perf.data ]
++ rc=12
```
[perf.data_202301260100.zip](https://github.com/multipath-tcp/mptcp_net-next/files/10507780/perf.data_202301260100.zip)
